File: tests/e2e/runtime/array_overrides/java/postgresql/src/main/java/com/example/db/queries/Queries.java
package com.example.db.queries;

import com.example.db.models.Record;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;

public final class Queries {
    private Queries() {}

    private static final String SQL_INSERT_RECORD = """
            INSERT INTO record (label, timestamps, uuids)
            VALUES (?, ?, ?);
            """;
    public static void insertRecord(Connection conn, String label, java.util.List<LocalDateTime> timestamps, java.util.List<UUID> uuids) throws SQLException {
        try (PreparedStatement ps = conn.prepareStatement(SQL_INSERT_RECORD)) {
            ps.setString(1, label);
            ps.setArray(2, conn.createArrayOf("timestamp", timestamps.toArray()));
            ps.setArray(3, conn.createArrayOf("uuid", uuids.toArray()));
            ps.executeUpdate();
        }
    }

    private static final String SQL_GET_RECORD = """
            SELECT id, label, timestamps, uuids
            FROM record
            WHERE id = ?;
            """;
    public static Optional<Record> getRecord(Connection conn, long id) throws SQLException {
        try (PreparedStatement ps = conn.prepareStatement(SQL_GET_RECORD)) {
            ps.setLong(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) return Optional.empty();
                return Optional.of(new Record(rs.getLong(1), rs.getString(2), java.util.Arrays.stream((Object[]) rs.getArray(3).getArray()).map(it -> ((java.sql.Timestamp) it).toLocalDateTime()).collect(java.util.stream.Collectors.toList()), java.util.Arrays.stream((Object[]) rs.getArray(4).getArray()).map(it -> (java.util.UUID) it).collect(java.util.stream.Collectors.toList())));
            }
        }
    }

    private static Boolean getNullableBoolean(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        boolean v = rs.getBoolean(col);
        return rs.wasNull() ? null : v;
    }
    private static Short getNullableShort(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        short v = rs.getShort(col);
        return rs.wasNull() ? null : v;
    }
    private static Integer getNullableInt(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        int v = rs.getInt(col);
        return rs.wasNull() ? null : v;
    }
    private static Long getNullableLong(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        long v = rs.getLong(col);
        return rs.wasNull() ? null : v;
    }
    private static Float getNullableFloat(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        float v = rs.getFloat(col);
        return rs.wasNull() ? null : v;
    }
    private static Double getNullableDouble(java.sql.ResultSet rs, int col) throws java.sql.SQLException {
        double v = rs.getDouble(col);
        return rs.wasNull() ? null : v;
    }
}

File: tests/e2e/runtime/array_overrides/kotlin/postgresql/src/main/kotlin/com/example/db/Queries.kt
package com.example.db

import java.sql.Connection
import java.time.LocalDateTime
import java.util.UUID

object Queries {

    private val SQL_INSERT_RECORD = """
        INSERT INTO record (label, timestamps, uuids)
        VALUES (?, ?, ?);
    """.trimIndent()
    fun insertRecord(conn: Connection, label: String, timestamps: List<LocalDateTime>, uuids: List<UUID>): Unit {
        conn.prepareStatement(SQL_INSERT_RECORD).use { ps ->
            ps.setString(1, label)
            ps.setArray(2, conn.createArrayOf("timestamp", timestamps.toTypedArray()))
            ps.setArray(3, conn.createArrayOf("uuid", uuids.toTypedArray()))
            ps.executeUpdate()
        }
    }

    private val SQL_GET_RECORD = """
        SELECT id, label, timestamps, uuids
        FROM record
        WHERE id = ?;
    """.trimIndent()
    fun getRecord(conn: Connection, id: Long): Record? {
        conn.prepareStatement(SQL_GET_RECORD).use { ps ->
            ps.setLong(1, id)
            ps.executeQuery().use { rs ->
                if (!rs.next()) return null
                return Record(rs.getLong(1), rs.getString(2), (rs.getArray(3).array as Array<*>).map { (it as java.sql.Timestamp).toLocalDateTime() }.toList(), (rs.getArray(4).array as Array<*>).map { (it as java.util.UUID) }.toList())
            }
        }
    }

    private fun getNullableBoolean(rs: java.sql.ResultSet, col: Int): Boolean? {
        val v = rs.getBoolean(col)
        return if (rs.wasNull()) null else v
    }
    private fun getNullableShort(rs: java.sql.ResultSet, col: Int): Short? {
        val v = rs.getShort(col)
        return if (rs.wasNull()) null else v
    }
    private fun getNullableInt(rs: java.sql.ResultSet, col: Int): Int? {
        val v = rs.getInt(col)
        return if (rs.wasNull()) null else v
    }
    private fun getNullableLong(rs: java.sql.ResultSet, col: Int): Long? {
        val v = rs.getLong(col)
        return if (rs.wasNull()) null else v
    }
    private fun getNullableFloat(rs: java.sql.ResultSet, col: Int): Float? {
        val v = rs.getFloat(col)
        return if (rs.wasNull()) null else v
    }
    private fun getNullableDouble(rs: java.sql.ResultSet, col: Int): Double? {
        val v = rs.getDouble(col)
        return if (rs.wasNull()) null else v
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> jdbcArrayToList(arr: java.sql.Array): List<T> =
        (arr.array as Array<T>).toList()
}
